Add an extra bounding box padding around a sub-app

andrsd opened the pull request:
<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

This allows adding extra bounding box around 1D meshes so that layered transfers can be done with 2D and 3D meshes.

andrsd committed:
Adding bounding box padding to MultiApp

This allows layered user objects to transfer values between 1D and 3D
meshes.

Closes #10966

andrsd commented:
@permcody This is a suggested solution for #10966. I do not have a test yet, but I am gonna do one today.

andrsd committed:
Adding a test

friedmud reviewed (changes requested):
I think this is backward.... this stuff is supposed to be in the _Transfer_.  Different Transfers will need different amounts of inflation / etc.

For instance: note that Transfers already have `displaced_source_mesh` and `displaced_target_mesh` parameters... having one here doesn't make any sense (it's not clear if you're forcing the sub-apps to _compute_ on the displaced mesh or what you're doing).

andrsd commented:
> it's not clear if you're forcing the sub-apps to compute on the displaced mesh or what you're doing

I think that `use_displaced_mesh = true` ([see the test file](https://github.com/idaholab/moose/pull/10967/commits/738fc0687ff519885b72e7497bc362efcc87e565#diff-65f18010e24decb9866a505f850c9779R92)) is pretty self-explanatory.

The problem is that the layered transfer does:
```
if (app_box.contains_point(*node))
```
where `app_box` for 1D meshes is something like (0, 0, 1)-(0, 0, 2) and good luck hitting that.

Moreover, `MultiApp` already has "bounding_box_inflation" parameter, so it made sense to be consistent and add the parameter there.

friedmud commented:
I still disagree about the displaced mesh.

You could have one Transfer where you want to use the displaced mesh and another where you don't.  It doesn't make sense to set it at the MultiApp level.

An app's "bounding box inflation" makes sense to me.  I think that parameter is currently correct.  I could also see that a Transfer might want some _extra_ inflation or something.

andrsd commented:
So the `use_displaced_mesh` parameter was not originally doing anything. The reason I made that do something was this line (in the transfer):
```
BoundingBox app_box = _multi_app->getBoundingBox(i);
```
This would give me the bounding box of an undisplaced mesh only, which was incorrect in my case.  What do you think about passing a parameter into this method that would determine if bounding box we are requesting should come from the displaced or undisplaced mesh? I agree that `use_displaced_mesh` parameter is a hack.

friedmud commented:
Yep - sounds good
On Mon, Mar 12, 2018 at 3:07 PM David Andrs <notifications@github.com>
wrote:

> So the use_displaced_mesh parameter was not originally doing anything.
> The reason I made that do something was this line (in the transfer):
>
> BoundingBox app_box = _multi_app->getBoundingBox(i);
>
> This would give me the bounding box of an undisplaced mesh only, which was
> incorrect in my case. What do you think about passing a parameter into this
> method that would determine if bounding box we are requesting should come
> from the displaced or undisplaced mesh? I agree that use_displaced_mesh
> parameter is a hack.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/10967#issuecomment-372342658>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA1JMdVMnP3gt-NDoPaWE82B9ttONfq8ks5tdo8ugaJpZM4Sg0MD>
> .
>

permcody commented:
Well you are assigned to his PR. You'll need to go in an review and merge
then.

On Mon, Mar 12, 2018 at 4:38 PM Derek Gaston <notifications@github.com>
wrote:

> Yep - sounds good
> On Mon, Mar 12, 2018 at 3:07 PM David Andrs <notifications@github.com>
> wrote:
>
> > So the use_displaced_mesh parameter was not originally doing anything.
> > The reason I made that do something was this line (in the transfer):
> >
> > BoundingBox app_box = _multi_app->getBoundingBox(i);
> >
> > This would give me the bounding box of an undisplaced mesh only, which
> was
> > incorrect in my case. What do you think about passing a parameter into
> this
> > method that would determine if bounding box we are requesting should come
> > from the displaced or undisplaced mesh? I agree that use_displaced_mesh
> > parameter is a hack.
> >
> > —
> > You are receiving this because you commented.
> >
> >
> > Reply to this email directly, view it on GitHub
> > <https://github.com/idaholab/moose/pull/10967#issuecomment-372342658>,
> or mute
> > the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AA1JMdVMnP3gt-NDoPaWE82B9ttONfq8ks5tdo8ugaJpZM4Sg0MD
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/10967#issuecomment-372376121>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AC5XILcTR5vNnMaSMP3mDmg1G_HO0IjLks5tdqR8gaJpZM4Sg0MD>
> .
>

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/163852/) on 738fc06 wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/10967/site)

This comment will be updated on new commits.

andrsd committed:
Fixing displaced_{source|target}_mesh parameters

Those parameters are set in the base class (MultiAppTransfer), so there
is no need to add them again in the child classes and initialize them
and possibly shadow the ones in the parent.

andrsd commented:
1. rebased on current devel
2. fixed the hack so that `MultiApp::getBoundingBox` takes and extra parameter
3. found some brain damage in `MultiAppTransfer` parameters, so fixed that too (see d73ad60).

friedmud commented on framework/src/multiapps/MultiApp.C:
```diff
@@ -360,21 +366,27 @@ MultiApp::restore()
   {
     Moose::ScopedCommSwapper swapper(_my_comm);
-    bbox = MeshTools::create_bounding_box(problem.mesh());
+    bbox = MeshTools::create_bounding_box(mesh);
```
One issue here: you are doing this every time for each Transfer.  I wonder how long this takes... I think it's a complete iteration over all the nodes in the mesh....

friedmud replied:
We usually store a bounding box on MooseMesh... can't we reuse that here?  (Maybe just get that one and inflate it slightly?)

andrsd replied:
If the mesh moved, we have to get the up-to-date bounding box. That's my understanding of this. 

The only bounding box on `MooseMesh` I see is the one returned by `MooseMesh::getInflatedProcessorBoundingBox`. Am I looking at the wrong place?  Also, this one should give back the global one, correct?

I can throw a patch on top of this to re-compute the BBox only if `displaced_mesh` is true and pre-cache the BBox the first time we call this.

friedmud replied:
Ah - you're right.

In that case - call `getInflatedBoundingBox()`... and cache the result when you can.

friedmud reviewed (commented):
Some questions on efficiency...

andrsd committed:
Do not recompute bounding box

If the sub-app's mesh does not move, we cache its bounding box, so we do
not have to recompute it all the time.  For moving meshes, we just have
to recompute to get the up-to-date boundaing box.
